### ISSUE TITLE
🛡 Security Audit 2026-04-20

### ISSUE THREAD

**Issue by fujihira-cloud** [FIshota/AI#1]
# 🛡 Daily Security Audit — 2026-04-20

## Summary
- pip-audit: **7 vulnerabilities** (CRITICAL: 1, HIGH: 3, MEDIUM: 2, NO-FIX: 1) in 3 packages
- outdated packages (system): **24** (major version jumps: 4)
- bandit: **22 issues** (HIGH: 0, MEDIUM: 22)
- secrets scan: **clean** — No hardcoded secrets detected

---

### pip-audit — PyTorch & Transformers

| CVE | Package | Installed | Fix Version | Severity |
|-----|---------|-----------|-------------|----------|
| **CVE-2025-32434** | torch | 2.4.1 | 2.6.0 | 🔴 CRITICAL |
| **CVE-2025-3730** | torch | 2.4.1 | 2.8.0 | 🔴 CRITICAL |
| **CVE-2024-48063** | torch | 2.4.1 | 2.5.0 | 🟠 HIGH |
| **CVE-2025-2953** | torch | 2.4.1 | 2.7.1rc1 | 🟠 HIGH |
| **CVE-2026-1839** | transformers | 4.57.6 | 5.0.0rc3 | 🟠 HIGH |
| **CVE-2025-69872** | diskcache | 5.6.3 | ⚠ no fix available | 🟡 MEDIUM |

**Notes:**
- **CVE-2025-32434** (torch): Arbitrary code execution via deserialization in `torch.load()`. Affects Python ≥3.10. Upgrade to ≥2.6.0.
- **CVE-2025-3730** (torch): RCE vector in distributed training module. Upgrade to ≥2.8.0 for full remediation.
- **CVE-2024-48063** (torch): Unsafe deserialization. Fixed in 2.5.0.
- **CVE-2025-2953** (torch): DoS/memory corruption in tensor operations. 2.7.1rc1 is the earliest fix (production build pending).
- **CVE-2026-1839** (transformers): Model loading vulnerability. Fix requires upgrading to 5.x — **breaking change** (review API compatibility before upgrading).
- **CVE-2025-69872** (diskcache): No upstream fix released yet. Monitor for patch or consider alternative caching strategy.

---

## ⚠ Outdated Dependencies

**Major version updates (security-relevant):**

| Package | Installed | Latest | Notes |
|---------|-----------|--------|-------|
| `cryptography` | 41.0.7 | 46.0.7 | ⚠ Security lib — 5 major versions behind |
| `setuptools` | 68.1.2 | 82.0.1 | Build tooling |
| `launchpadlib` | 1.11.0 | 2.1.0 | API breaking change |
| `wadllib` | 1.3.6 | 2.0.0 | API breaking change |

Minor/patch updates: 20 packages (argcomplete, blinker, charset-normalizer, conan, dbus-python, httplib2, lazr.uri, oauthlib, packaging, patch-ng, pip, PyGObject, PyJWT, pyparsing, PyYAML, six, wheel, xmltodict, yq)

> ⚠ Note: `cryptography 41.0.7` (system) is severely outdated. The project `requirements.txt` correctly pins `cryptography>=44.0.0` — ensure Docker/deployment environments use the constrained version.

---

## 📋 Bandit Findings (Medium only — no High issues detected)

**B601 — Paramiko shell injection risk** (6 findings)

| File | Line | Issue |
|------|------|-------|
| `core/server_home.py` | 204 | `exec_command("echo ok")` — static string, low risk |
| `core/server_home.py` | 228 | `exec_command(cmd)` — **dynamic `cmd` variable — review input sanitization** |
| `core/server_home.py` | 261 | `exec_command("uptime -p")` — static string |
| `core/server_home.py` | 265 | `exec_command("df -h / \| tail -1")` — static string |
| `core/server_home.py` | 269 | `exec_command("free -h \| grep Mem \| awk ...")` — static string |
| `core/server_home.py` | 275 | `exec_command("docker ps ...")` — static string |

> Line 228 is the only truly dynamic call — verify `cmd` is not constructed from user input.

**B608 — SQL injection (string-based query construction)** (4 findings)

| File | Line | Issue |
|------|------|-------|
| `core/data_exporter.py` | 160 | `f"SELECT * FROM {table_name}"` — table name injected directly |
| `core/memory.py` | 645 | f-string WHERE clause construction |
| `core/memory.py` | 667 | f-string WHERE clause with parameterized values (low risk) |
| `core/memory_compressor.py` | 68 | f-string UPDATE with parameterized IN clause (low risk) |

> `data_exporter.py:160` is highest risk — `table_name` should be validated against an allowlist of known table names.

**B310 — urllib.request.urlopen scheme audit** (11 findings)

Files: `core/competitor_analyzer.py:116`, `core/github_learner.py:180`, `core/image_gen.py:156`, `core/research_agent.py:190`, `core/tts/engine.py:126,145,165`, `core/web_fetcher.py:36,56,108,171`

> Most have `# noqa: S310` suppression or only use HTTP/HTTPS. `core/research_agent.py:190` lacks a scheme assertion — add URL validation before calling.

**B314 — Unsafe XML parsing** (1 finding)

| File | Line | Issue |
|------|------|-------|
| `core/web_fetcher.py` | 57 | `ET.parse(resp)` on external HTTP response — vulnerable to XXE |

> Replace with `defusedxml.ElementTree.parse()` or call `defusedxml.defuse_stdlib()` at startup.

---

## 🔍 Secret Scan

```
No hardcoded secrets detected
```

Patterns scanned: `sk-*` (OpenAI keys), `ghp_*` (GitHub PATs), `AKIA*` (AWS keys), PEM private keys.
Excluded: `tests/`, `docs/`, `.git/`

---

## Recommendations (Priority Order)

1. **[CRITICAL] Upgrade `torch` to ≥2.6.0 immediately** — CVE-2025-32434 is an RCE via deserialization. Run `pip install "torch>=2.6.0"`. For Py<3.13 environments, update `requirements.txt` constraint from `<2.5` to `<2.6` and test compatibility. Full remediation of CVE-2025-3730 requires 2.8.0 (monitor stable release).

2. **[HIGH] Evaluate `transformers` upgrade path to 5.x** — CVE-2026-1839 is only fixed in 5.0.0rc3+. Audit breaking API changes in transformers 5.x and create a migration branch. Pin to `<5.0` with a documented exception until tested.

3. **[MEDIUM] Fix `diskcache` exposure (CVE-2025-69872)** — No upstream fix. Assess if diskcache is used in an adversarial context (e.g., shared cache with untrusted data). Consider wrapping cache values with integrity checks or switching to `cachetools` for untrusted data paths.

4. **[MEDIUM] Harden `core/data_exporter.py:160` SQL** — Validate `table_name` against an explicit allowlist (e.g., `ALLOWED_TABLES = {"memories", "logs", ...}`) before interpolating into the query string.

5. **[MEDIUM] Replace `xml.etree.ElementTree` with `defusedxml` in `core/web_fetcher.py:57`** — External RSS/XML feeds can contain malicious entity expansions (XXE). One-line fix: `import defusedxml.ElementTree as ET`.
